feature_tests/cpp2/include/Bar.d.hpp

```diff
@@ -21,10 +22,10 @@ class Bar {
 
   inline const Foo& foo() const;
 
-  inline const capi::Bar* AsFFI() const;
-  inline capi::Bar* AsFFI();
-  inline static const Bar* FromFFI(const capi::Bar* ptr);
-  inline static Bar* FromFFI(capi::Bar* ptr);
+  inline const ::capi::Bar* AsFFI() const;
```

Comment:
This comment seems to keep getting buried:

I don't think we _need_ the `::` though? Resorting to `::` seems like a lazy solution to a problem that should be better solved by the code keeping track of what namespace the types are going in. This is a code hygiene comment separate from my other concern about avoiding the use of the global namespace.

Reply:
I think if you just remove the `::`, things should "just work" based on C++ name resolution.

Reply:
> Resorting to `::` seems like a lazy solution to a problem that should be better solved by the code keeping track of what namespace the types are going in

I don't agree that it's being lazy. Tracking namespaces is completely unnecessary here. Unnecessary `::capi` is a code smell in _manually written_ code because it can confuse the reader ("aren't we already in `capi`?") and is unnecessary work. This isn't code that is going to be manually edited, and it's saving us unnecessary work here.

What is the concrete argument for avoiding `::`? It's only "lazy" if there's some concrete benefit we're avoiding. I have yet to be presented with what the other side of that tradeoff looks like.

Reply:
Did you try just removing the `::`? You don't use the global namespace qualifier when there is a namespace involved (in your `icu4x::` and `ns::` examples) so you shouldn't need it here, either.

Reply:
Yes, you get a failure

```
./tests/../include/AttrOpaque1Renamed.hpp:27:103: error: ‘Unnamespaced’ in namespace ‘ns::capi’ does not name a type
   27 |     void namespace_AttrOpaque1_use_unnamespaced(const ns::capi::AttrOpaque1Renamed* self, const capi::Unnamespaced* _un);
      |                                                                                                       ^~~~~~~~~~~~
./tests/../include/AttrOpaque1Renamed.hpp: In member function ‘void ns::AttrOpaque1Renamed::use_unnamespaced(const Unnamespaced&) const’:
./tests/../include/AttrOpaque1Renamed.hpp:54:14: error: cannot convert ‘const capi::Unnamespaced*’ to ‘const int*’
   54 |     _un.AsFFI());
      |     ~~~~~~~~~^~
      |              |
      |              const capi::Unnamespaced*
```

Reply:
We have multiple `capi` namespaces, one in each C++ namespace. The goal of this PR is so that there is no polluted global `capi` namespace for people who are namespacing everything. Which means that for every C++ namespace we define, there is a `capi` namespace under it. For ICU4X, that just means one `icu4x::capi` namespace, but projects using multiple namespaces get multiple.

(There's still one for `diplomat_runtime` stuff, I plan to address that separately)

Reply:
But the types are only in a namespace, as far as I see, if there is a namespace defined in the bridge.

For example, in this file:

```cpp
#ifndef BorrowedFields_D_HPP
#define BorrowedFields_D_HPP
#include <stdio.h>

// ...

namespace capi {typedef struct Bar Bar; }
class Bar;

namespace capi {
    typedef struct BorrowedFields {
      DiplomatString16View a;
      // ...
    } BorrowedFields;
}
struct BorrowedFields {
  std::u16string_view a;
  // ...

  inline static diplomat::result<BorrowedFields, diplomat::Utf8Error> from_bar_and_strings(const Bar& bar, std::u16string_view dstr16, std::string_view utf8_str);

  inline ::capi::BorrowedFields AsFFI() const;
  inline static BorrowedFields FromFFI(::capi::BorrowedFields c_struct);
};
```

I don't see why you need `::capi` instead of just `capi` in that file.

But OK, I see now when you have mixed namespace and non-namespace why you need it

```cpp
void namespace_AttrOpaque1_use_unnamespaced(const ns::capi::AttrOpaque1Renamed* self, const capi::Unnamespaced* _un);
```

The `capi` it finds will be `::ns::capi` when we want it to instead find `::capi`.

Reply:
That's an interesting point, though. `capi` is not a namespace that we might want on the top level; it is a namespace that is not Diplomat or ICU4X specific. It's fine when it is inside of the client's chosen namespace, but putting a `capi` namespace on the top level might not be desirable.

Reply:
> I don't see why you need ::capi instead of just capi in that file.

Yes, because the _generating code_ is much simpler and less likely to be incorrect if it does not attempt to polish this.

Complicating the C backend's `fmt_type_name()` to be context-aware in this way is tricky and very likely to be broken in subtle-but-untested-ways, _especially_ since there is already a bit of a C-vs-C++ divergence in C backend code, and it's not worth it.

You're right that the `::capi` is unnecessary there, from the point of view of a file that was written manually, where "remove the `::`" is a task that is just as simple as it sounds. But it wasn't written manually. That is not a straightforward task, and nothing so far has been said to indicate that doing so will be at all worth the complexity and brittleness.